cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/AuthorizationArtifactRepository.java

```diff
@@ -0,0 +1,299 @@
+  }
+
+  @Override
+  public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile) throws Exception {
+   return addArtifact(artifactId, artifactFile, null, null);
+  }
+
+  @Override
+  public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
+                                    @Nullable Set<ArtifactRange> parentArtifacts,
+                                    @Nullable Set<PluginClass> additionalPlugins) throws Exception {
+    return addArtifact(artifactId, artifactFile, parentArtifacts, additionalPlugins,
+                       Collections.<String, String>emptyMap());
+  }
+
+  @Override
+  public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
+                                    @Nullable Set<ArtifactRange> parentArtifacts,
+                                    @Nullable Set<PluginClass> additionalPlugins,
+                                    Map<String, String> properties) throws Exception {
+    // To add an artifact, a user must have ADMIN privilege on the artifact is being added
+    // This method is used to add user app artifacts, so enforce authorization on the specified, non-system namespace
+    Principal principal = authenticationContext.getPrincipal();
+    authorizationEnforcer.enforce(artifactId.toEntityId(), principal, Action.ADMIN);
+    return delegate.addArtifact(artifactId, artifactFile, parentArtifacts, additionalPlugins, properties);
+  }
+
+  @Override
+  public void writeArtifactProperties(Id.Artifact artifactId, Map<String, String> properties) throws Exception {
+    authorizationEnforcer.enforce(artifactId.toEntityId(), authenticationContext.getPrincipal(), Action.ADMIN);
+    delegate.writeArtifactProperties(artifactId, properties);
+  }
+
+  @Override
+  public void writeArtifactProperty(Id.Artifact artifactId, String key, String value) throws Exception {
+    authorizationEnforcer.enforce(artifactId.toEntityId(), authenticationContext.getPrincipal(), Action.ADMIN);
+    delegate.writeArtifactProperty(artifactId, key, value);
+  }
+
+  @Override
+  public void deleteArtifactProperty(Id.Artifact artifactId, String key) throws Exception {
+    authorizationEnforcer.enforce(artifactId.toEntityId(), authenticationContext.getPrincipal(), Action.ADMIN);
+    delegate.deleteArtifactProperty(artifactId, key);
+  }
+
+  @Override
+  public void deleteArtifactProperties(Id.Artifact artifactId) throws Exception {
+    authorizationEnforcer.enforce(artifactId.toEntityId(), authenticationContext.getPrincipal(), Action.ADMIN);
+    delegate.deleteArtifactProperties(artifactId);
+  }
+
+  @Override
+  public void addSystemArtifacts() throws Exception {
+    // to add system artifacts, users should have write privileges on the system namespace
+    Principal principal = authenticationContext.getPrincipal();
+    authorizationEnforcer.enforce(NamespaceId.SYSTEM, principal, Action.ADMIN);
```

Comment:
Yeah, we decided to change it from WRITE to ADMIN to refresh the system artifacts